synapse/storage/databases/main/deviceinbox.py

```diff
@@ -891,6 +892,46 @@ def _add_messages_to_local_device_inbox_txn(
                 ],
             )
 
+    async def delete_device_message(self, stream_id: int) -> bool:
+        """Delete a specific device message from the message inbox.
+
+        Args:
+            stream_id: the stream ID identifying the message.
+        Returns:
+            True if the message has been deleted, False if it didn't exist.
+        """
+        try:
+            await self.db_pool.simple_delete_one(
+                "device_inbox",
+                keyvalues={"stream_id": stream_id},
+                desc="delete_device_message",
+            )
+        except StoreError:
+            # Deletion failed because device message does not exist
+            return False
+        return True
+
+    async def get_all_device_messages(
```

Comment:
Is there any concern about performance with this function if the device inbox is very large? Presumably we can't just grab the `RoomKeyRequest`s because that information is locked away in the `Json` blobs?

Reply:
For the perf I am not sure. `get_messages_for_user_devices` is kind of unbounded too.

I can narrow down the set of keys by using a `LIKE` with the key type on the JSON blob. Is it worth it, I am not sure.